src/libcrun/sha256.c

```diff
@@ -0,0 +1,483 @@
+/* sha256.c - Functions to compute SHA256 and SHA224 message digest of files or
+   memory blocks according to the NIST specification FIPS-180-2.
+
+   Copyright (C) 2005-2006, 2008-2022 Free Software Foundation, Inc.
+
+   This file is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Lesser General Public License as
+   published by the Free Software Foundation; either version 2.1 of the
+   License, or (at your option) any later version.
+
+   This file is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.  */
+
+/* Written by David Madore, considerably copypasting from
+   Scott G. Miller's sha1.c
+*/
+
+#include <config.h>
+
+/* Specification.  */
+#if HAVE_OPENSSL_SHA256
```

Comment:
Is static compilation an option?

Reply:
That would turn every crypto library security update into a crun security update.

(I’d expect the overhead to be in FIPS status checks and self-integrity checking more than just library size.

Reply:
I was not sure whether adding more code or using libgcrypt, so I am glad you commented here!

I've just pushed a new version where libgcrypt is linked dynamically

Reply:
I think this is only being used to create `key` for the cache, from looking at the code I see no intention of using it for any other task.

I think using library and static linking will just increase the object size.

Reply:
> I was not sure whether adding more code or using libgcrypt, so I am glad you commented here!
> 
> I've just pushed a new version where libgcrypt is linked dynamically

@giuseppe @mtrmac just a small doubt and I could be wrong , isn't this going to load entire library at the beginning of the program ? plus it just adds to dependency.

If `sha256` is just being used to calculate checksum of seccomp content for uniquely identifying and storing cache then isn't maintaing small part of `sha256` compute with the crun code better **for performance** ? ( since security is not a concern here even if code gets outdated ).

Again above is just a question to clarify to my doubts :)

Reply:
It might well be the case that this, just using it for a hash key from trusted (?) data, is not even considered cryptography. I didn't analyze the code or the security assumptions in detail, against any set of requirements.

But even if that were true, it’s just easier to not have the code around and not to be asked to justify this.

Also once the code is included, it might get started to be used for other purposes.

Reply:
> @giuseppe @mtrmac just a small doubt and I could be wrong , isn't this going to load entire library at the beginning of the program ? plus it just adds to dependency.

- “Loading” a dynamically-linked library is just a `mmap`; that has the performance cost of a few syscalls, regardless of library size. And if that library is used in any other process, that might not hit the disk at all — while a statically-linked implementation increases the size of crun itself, which could require a larger disk read if crun is not in memory already.
- There is some cost in dynamically linking, that primarily depends on the number of variable references (not code references) in the library
- A library then might run some initialization code; that _might_ be non-trivial, but that’s also usually because it’s valuable.

If we are doing all of this for performance, it’s certainly worth measuring to avoid surprises.

Without measurement, it’s really anyone’s guess — library cost vs. the cost of using an inefficient portable implementation instead of things like https://en.wikipedia.org/wiki/Intel_SHA_extensions .

OTOH the packaging / review overhead of having to discuss this copy&pasted code is known, and forever.

Reply:
@mtrmac ah thanks for the detailed explanation. SGTM.

My take in previous comment was to port only small parts of `sha256` into `crun`'s src maybe like: https://github.com/redis/redis/blob/unstable/src/sha256.c ( based on assumption that it is not being used for any security purpose )

But I agree that using a library adds lot of other benefits + (it has inbuilt hardware optimizations) with no maintenance and all these things easily outweighs maintaining custom code.

Plus I am not sure how to benchmark memory usage of dynamically loading `libgcrypt` since most likely its already loaded by other process on my os, which is very likely true for most of the real use-case as well. ( @giuseppe any hints how can i benchmark this ? )

Thanks again. :)

Reply:
re static liking:  compilers are generally smart enough to erase dead code.  So static linking does not equate to the an increase of the size of all objects of the library.

re benchmarks:
* compare the sizes of statically linked with dynamically linked one
* compare performance of the two with hot and cold caches
* for memory usage, you can use valgrind

Reply:
at least on Fedora, we are already linking to libgcrypt through libsystemd, so it won't really affect us.

```
$ ldd /lib64/libsystemd.so.0
	linux-vdso.so.1 (0x00007ffd1cf64000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007f035d5b3000)
	libzstd.so.1 => /lib64/libzstd.so.1 (0x00007f035d500000)
	liblz4.so.1 => /lib64/liblz4.so.1 (0x00007f035d4dc000)
	libcap.so.2 => /lib64/libcap.so.2 (0x00007f035d4d2000)
	libgcrypt.so.20 => /lib64/libgcrypt.so.20 (0x00007f035d395000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f035d375000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f035d000000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f035d6df000)
	libgpg-error.so.0 => /lib64/libgpg-error.so.0 (0x00007f035d34d000)

$ ldd /usr/bin/crun 
	linux-vdso.so.1 (0x00007ffd2439f000)
	libcriu.so.2 => /lib64/libcriu.so.2 (0x00007f18a8ec8000)
	libsystemd.so.0 => /lib64/libsystemd.so.0 (0x00007f18a8deb000)
	libseccomp.so.2 => /lib64/libseccomp.so.2 (0x00007f18a8dcb000)
	libcap.so.2 => /lib64/libcap.so.2 (0x00007f18a8dc1000)
	libyajl.so.2 => /lib64/libyajl.so.2 (0x00007f18a8db5000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f18a8d95000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f18a8a00000)
	libprotobuf-c.so.1 => /lib64/libprotobuf-c.so.1 (0x00007f18a8d87000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007f18a8d5c000)
	libzstd.so.1 => /lib64/libzstd.so.1 (0x00007f18a8ca9000)
	liblz4.so.1 => /lib64/liblz4.so.1 (0x00007f18a8c85000)
	libgcrypt.so.20 => /lib64/libgcrypt.so.20 (0x00007f18a88c3000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f18a8f67000)
	libgpg-error.so.0 => /lib64/libgpg-error.so.0 (0x00007f18a8c5d000)
```